### PR TITLE
RMSNorm, Normformer and fixes from merging 1.8.0 into main

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Implements the following

1. RMSNorm normalization, used in T5 when not using bias terms.
2. Normformer that adds extra normalization layers after self-attention and between the MLP feedforward layers. Also adds the option of configuring the additional MHA scale parameter.
3. Fixed some issues with merging code from 1.8.0 to main that removed code relevant to removing bias terms.

**Collection**: NLP

# Changelog 
- Mostly changes to `transformer.py`

# Usage
* You can potentially add a usage example below

Configure the following args in your T5/BART yaml file

encoder_arch: "normformer"
decoder_arch: "normformer"
headscale: True
normalization: rmsnorm
bias: False

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — DISMISSED** [review]
I'll just give a cursory review. It seems to be fine, there may be a need to think about refactoring out the core of Megatron into a base classz and have seperate classes implement such functionality instead. Right now, it's going down the Jasper block path and soon it will become untenable to maintain c

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 1]
```diff
@@ -276,6 +298,12 @@ def __init__(
         self.hidden_size_per_attention_head = safe_divide(projection_size, num_attention_heads)
         self.num_attention_heads_per_partition = safe_divide(num_attention_heads, world_size)
 
+        # Headscale
+        if headscale:
+            self.head_scale_tensor = torch.nn.Parameter(
+                torch.ones(1, self.num_attention_heads_per_partition, 1, self.hidden_size_per_attention_head)
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:304]: Should the scaling attention be a scalar per attention head and not per hidden unit in each head?

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 2]
```diff
@@ -113,10 +114,19 @@ def __init__(
         onnx_safe=False,
         activation='gelu',
         bias=True,
+        arch='transformer',
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:117]: We should add some hierarchy to these methods' arguments otherwise it will soon become unmanageable.  E.g. group feature related arguments into a dictionary objects.

**Review by yidong72 — COMMENTED** [review]
Just one comment that we should think of grouping the arguments into some hierarchy. A flatten list of arguments are very hard to follow and extend.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 3]
```diff
@@ -732,64 +845,77 @@ def forward(
...
-            if self.training:
-                bias_dropout_add_func = bias_dropout_add_fused_train
-            else:
-                bias_dropout_add_func = bias_dropout_add_fused_inference
-        elif self.bias and not self.bias_dropout_fusion:
-            bias_dropout_add_func = get_bias_dropout_add(self.training)
-        else:
-            bias_dropout_add_func = dropout_add
+
+        bias_dropout_add_func = self._get_bias_droput_add_func(arch=self.arch, position_after='attention')
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:866]: In the Normformer paper I could not find the dropout after attention. Do we add this here in purpose? (or perhaps I missed it in the paper?)

**Review by michalivne — COMMENTED** [review]
Great PR! I added a few comments/questions.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:304]: in the paper, it seems to be per attention head
image

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 4]
```diff
@@ -557,6 +587,9 @@ def forward(
         # change view [b, np, sq, hn]
         context_layer = context_layer.view(*output_size)
 
+        if self.headscale:
+            context_layer = context_layer * self.head_scale_tensor
+
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:592]: Missing a `LN` after the `HeadScaleMHA` ?
image

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 5]
```diff
@@ -732,64 +845,77 @@ def forward(
...
+            normalization_output = self.post_inter_attention_normalization(layernorm_input)
 
         # MLP.
-        mlp_output, mlp_bias = self.mlp(layernorm_output)
+        mlp_output, mlp_bias = self.mlp(normalization_output)
 
         # Second residual connection.
         if self.apply_residual_connection_post_layernorm:
-            residual = layernorm_output
+            residual = normalization_output
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:910]: `residual` for FFN should be scaled by a learnable parameter `lambda_resid`?
image

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 4 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:592]: It happens outside the ParallelAttention layer in ParallelTransformerLayer_.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 5 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:910]: The paper shows that it hurts perplexity so I didn't add it (see Fig.2 in their paper)

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:304]: Yeah I missed the "scalar" part here. Since the equation is ambiguous about whether gamma is a scalar or tensor. Will fix it.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 3 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:866]: The normformer paper doesn't really include any modifications to where dropout is performed so this part is exactly how it is in the original transformer arch.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks for the changes!
